### ISSUE TITLE
Add units to topologyobjects

### ISSUE THREAD

**Issue by swails** [ParmEd/ParmEd#492]
ParmEd has, for some time, understood units and been able to deal with dimensional analysis and automatic unit conversion.  However, all of the parameter types in `topologyobjects.py` do not expose their parameters with units (they accept parameters with units and throw errors for incompatible units).  This has led to some confusion.

I would like to investigate the implications of, and consider implementing, the infrastructure to always supply units to the parameter types.  Possible downsides are severe API breakage and possibly severe performance degradation (as the dimensional analysis can be slow).

**Comment by jchodera** [comment]
xref: https://github.com/choderalab/perses/issues/55

There a few proposal ideas floating around:
* 1. Convert everything to unit-bearing quantities, possibly also to OpenMM-style representations of force constants
* 2. Use Hungarian-like notation to extend the API without breaking things:
```python
>>> bt = BondType(100.0, 1.0)
>>> bt.k
100.0
>>> bt.req
1.0
>>> bt.uk
Quantity(value=100, unit=kilocalorie/(angstrom**2*mole))
>>> bt.ureq
Quantity(value=1.0, unit=angstrom)
```
* 3. Have subclasses that can hold unit-bearing quantities
* 4. Add accessors to `topologyobjects` that get quantities in different representations

Maybe we should start by figuring out how bad API breakage would be.  Is it a non-starter, or is it something that could be tolerated?

**Comment by pgrinaway** [comment]
It seems like 2 would be least disruptive, and I certainly would be fine with it, but it does create some API clutter...

**Comment by swails** [comment]
3 wouldn't really work because. 2 is going to be the easiest and safest one to get working soonest. 

1 would be nicest in retrospect, but will take a while to even see if it's viable (that is, it could take a week just to determine it won't work). I don't see how 4 differs from 2. 

My vote would be for 2 the more I think about it.

**Comment by swails** [comment]
Because you can't insert the needed subclasses where they are needed in a good way. 

Typing from my phone is hard :)

**Comment by swails** [comment]
Is there still a need for this?  It hasn't cropped up in well over a year...

**Comment by jchodera** [comment]
Is there still a need for preventing mistakes by using units? Yes!

**Comment by jchodera** [comment]
The amount of lost human work time accumulated so far due to these unit issues is at least weeks.

How much effort do you think it would be to add this feature? (I think the consensus was that option 2 was optimal?)

**Comment by swails** [comment]
This is implemented in #864.  The one class that does not have a `u<parameter>` property is the CMAP grid, since there wasn't really a straightforward way of making one there.  It will now convert units if provided in the input, though.
